compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java

```diff
@@ -197,6 +234,139 @@ private void performCodeGen(boolean shrink) {
+            // Its immediate dependent modules are marked as "used" and they are optimized after that.
+            // This process happens till all "used" modules are exhausted.
+            if (shouldOptimize(moduleContext) && (isRootModule(moduleContext) || moduleContext.isUsed())) {
+                this.usedBIRNodeAnalyzer.analyze(moduleContext.bLangPackage());
+                updateNativeDependencyMap(moduleContext);
+            }
+        }
+
+        CodeGenOptimizationReportEmitter.emitBirOptimizationDuration();
+
+        if (this.packageContext.project().buildOptions().verbose()) {
+            CodeGenOptimizationReportEmitter.emitCodegenOptimizationReport(
+                    this.usedBIRNodeAnalyzer.pkgWiseInvocationData, getOptimizationReportPath(),
+                    packageContext.project().kind());
+        }
+    }
+
+    private Path getOptimizationReportPath() {
+        return this.packageContext.project().sourceRoot;
+    }
+
+    protected Set<PackageID> getOptimizedPackageIDs() {
+        return this.usedBIRNodeAnalyzer.pkgWiseInvocationData.keySet();
+    }
+
+    private void markTestDependenciesForDuplicateBIRGen() {
+        for (int i = pkgResolution.topologicallySortedModuleList().size() - 1; i >= 0; i--) {
+            ModuleContext moduleContext = pkgResolution.topologicallySortedModuleList().get(i);
+            BLangPackage bLangPackage = moduleContext.bLangPackage();
+
+            if (isRootModule(moduleContext) && bLangPackage.hasTestablePackage()) {
+                Set<BPackageSymbol> buildPkgDependencies = new HashSet<>();
+                Set<BPackageSymbol> testablePkgDependencies = new HashSet<>();
+                collectDependencies(bLangPackage.symbol, buildPkgDependencies);
+                collectDependencies(bLangPackage.getTestablePkg().symbol, testablePkgDependencies);
+
+                buildPkgDependencies.stream()
+                        .filter(testablePkgDependencies::contains).filter(this::isNotWhiteListedPkg)
+                        .forEach(pkgSymbol -> {
+                            pkgSymbol.shouldGenerateDuplicateBIR = true;
+                            // Have to use a hashmap because the pkgIds get mutated later
+                            JvmCodeGenUtil.duplicatePkgsMap.put(
+                                    pkgSymbol.pkgID.orgName + pkgSymbol.pkgID.getNameComps().toString(),
+                                    pkgSymbol.pkgID);
+                        });
+                return;
+            }
+        }
+    }
+
+    private boolean isNotWhiteListedPkg(BPackageSymbol pkgSymbol) {
+        HashSet<String> whiteListedPkgNames =
+                new HashSet<>(Arrays.asList("ballerina/observe", "ballerina/jballerina", "ballerina/lang"));
+        for (String pkgName : whiteListedPkgNames) {
+            if (pkgSymbol.pkgID.toString().contains(pkgName)) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
Does `pkgSymbol.pkgID.toString()` contain an additional part in the string? is that a variable part for time to time?

If that's not the case, we can use.
```suggestion
  return  whiteListedPkgNames.contains(pkgSymbol.pkgID.toString())
```

Reply:
Fixed in 6c2a0bb52a3a5a2cfdab4dbcd5410f1f7cd65b36. Now we use only one loop to validate the condition.